src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,824 @@
+ *
+ * @return Name of the group.
+ */
+static const char *
+get_groupname(void)
+{
+    gid_t egid;
+    struct group *gr;
+
+    pthread_mutex_lock(&commit_ctx.lock);
+    egid = getegid();
+    pthread_mutex_unlock(&commit_ctx.lock);
+
+    gr = getgrgid(egid);
+    if (!gr) {
+        return NULL;
+    }
+    return gr->gr_name;
+}
+
+/**
+ * @brief Get effective user name.
+ *
+ * It might not be safe to call geteuid() is called in a different context than this file
+ *
+ * @return Name of the user.
+ */
+static const char *
+get_username(void)
+{
+    uid_t euid;
+    struct passwd *pw;
+
+    pthread_mutex_lock(&commit_ctx.lock);
+    euid = geteuid();
+    pthread_mutex_unlock(&commit_ctx.lock);
+
+    pw = getpwuid(euid);
+    if (!pw) {
+        return NULL;
+    }
+    return pw->pw_name;
+}
+
+/**
+ * @brief Check if user can read or write with current permissions.
+ *
+ * @param[in] user User name to check.
+ * @param[in] group Group name to check.
+ * @param[in] perm Permissions on the module.
+ * @param[in] check The type of permission to check.
+ * @return 0 When can read.
+ * @return 1 When cannot read.
+ * @return -1 When error encountered.
+ */
+static int
+check_permissions(char *user, char *group, mode_t perm, perm_t check)
+{
+    const char *c_user, *c_group;
+    mode_t needed;
+
+    c_user = get_username();
+    c_group = get_groupname();
+    if (!c_user | !c_group) {
+        return -1;
+    }
+
+    /* If other can access no need to check further */
+    if (check == R) {
+        needed = S_IROTH;
+    } else {
+        needed = S_IWOTH;
+    }
+    if (perm & needed) {
+        return 0;
+    }
+
+    /* If group allowed and matching */
+    if (check == R) {
+        needed = S_IRGRP;
+    } else {
+        needed = S_IWGRP;
+    }
+    if (!strcmp(group, c_group) && perm & needed) {
+        return 0;
+    }
+
+    /* If user allowed and matching */
+    if (check == R) {
+        needed = S_IRUSR;
+    } else {
+        needed = S_IWUSR;
+    }
+    if (!strcmp(user, c_user) && perm & needed) {
+        return 0;
+    }
+
+    return 1;
+}
+
+/**
+ * @brief Restore running using the backup files.
+ * Thread run after the timer in commit_ctx_s runs out.
+ */
+static void
+changes_rollback(union sigval UNUSED(sev))
+{
+    struct lyd_node *node = NULL;
+    const struct ly_ctx *ctx = NULL;
+    struct lys_module *module = NULL;
+    char *user = NULL, *group = NULL, *path = NULL, *module_name = NULL;
+    mode_t perm;
+    DIR *dir = NULL;
+    struct dirent *dirent = NULL;
+
+    VRB("Confirmed commit timeout reached. Restoring original running.");
+    if (sr_session_switch_ds(np2srv.sr_sess, SR_DS_RUNNING)) {
+        ERR("Failed switching session.");
+        goto cleanup;
+    }
+    ctx = sr_get_context(np2srv.sr_conn);
+
+    /* Iterate over all files in backup directory */
+    dir = opendir(NP2SRV_DIR);
+    if (!dir) {
+        ERR("Failed opening netopeer2 server directory '%s'", NP2SRV_DIR);
```

Comment:
Generally, please use `"` instead of `'`, that is what is used for all quotations.